clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc

```diff
@@ -4771,6 +4962,122 @@ sycl::event engine_ext::async_rnn_backward(
   return e;
 }
 // DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|get_dropout_state_size|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|get_dropout_state_size_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline
+size_t engine_ext::get_dropout_state_size(){
+  return _random_engine_state_size;
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|get_dropout_workspace_size|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|get_dropout_workspace_size_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline size_t
+engine_ext::get_dropout_workspace_size(const memory_desc_ext &src_desc) {
+  return src_desc.get_size();
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|async_dropout_forward|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|async_dropout_forward_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline
+sycl::event engine_ext::async_dropout_forward(dropout_desc &desc,
+                                              const memory_desc_ext &src_desc,
+                                              void *src,
+                                              const memory_desc_ext &dst_desc,
+                                              void *dst, void *workspace,
+                                              size_t workspace_size) {
+  if (workspace_size < src_desc.get_size()) {
+    throw std::runtime_error("async_dropout_forward: no sufficient workspace.");
+  }
+  float p = desc.get_probability();
+  if (p == 1.f) {
+    return _q->memset(dst, 0, dst_desc.get_size());
+  } else if (p == 0.f) {
+    return async_reorder(1.f, src_desc, src, 0.f, dst_desc, dst);
+  }
```

Comment:
Are you sure of the order here? I would expect the if conditions to be swapped as
in general a Bernoulli distribution returns 1 with probability p. (see [onemkl doc](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/oneapi-mkl-rng-bernoulli.html#GUID-7E457D80-1FEA-4E4E-B4C9-C3AFA7490A99)).
So if p == 1, the mask is all ones so the src data should be returned.
if p == 0, the mask is all zeros so dst is set to 0.

Reply:
Yes, I'm sure. For cuDNN, the value of dropout means the probability with which the value from input is set to zero during the dropout layer. So, the p in cuDNN is equivalent to (1 - p) in cuDNN.